google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,205 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "monitor"
+require "concurrent"
+require "google/cloud/pubsub/publish_result"
+require "google/cloud/pubsub/service"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # @private
+      # # Publisher
+      #
+      class Publisher
+        include MonitorMixin
+
+        MAX_BYTES = 5*1024*1024
+        MAX_MESSAGES = 1000
+        TIMER_DELAY = 0.25
+
+        attr_reader :service, :queue
+        attr_reader :max_bytes, :max_messages, :timer_delay
+        attr_reader :thread_pool, :timer_task
+
+        def initialize service, max_bytes: MAX_BYTES,
+                       max_messages: MAX_MESSAGES, timer_delay: TIMER_DELAY
+          @service = service
+          @max_bytes = max_bytes
+          @max_messages = max_messages
+          @timer_delay = timer_delay
+
+          @queue = {}
+          @thread_pool = Concurrent::FixedThreadPool.new(
+            [2, Concurrent.processor_count * 2].max,
+            fallback_policy: :caller_runs
```

Comment:
This `fallback_policy` seems dangerous -- it means that a function that advertises itself as async can occasionally become blocking.

Why not leave `max_queue` unbounded or set `fallback_policy` to `:abort`?